### ISSUE TITLE
ValueError: Only scalar actions are supported now!!

### ISSUE THREAD

**Issue by shamim237** [tensorflow/agents#665]
I am getting this error while trying to create a custom environment and feed it to the DQN agent. 

**ValueError: Only scalar actions are supported now, but action spec is: BoundedTensorSpec(shape=(1,), dtype=tf.float32, name='action', minimum=array(0., dtype=float32), maximum=array(1., dtype=float32))
  In call to configurable 'DqnAgent' (<class 'tf_agents.agents.dqn.dqn_agent.DqnAgent'>)**

Here is my action_spec: 

**self._action_spec = array_spec.BoundedArraySpec(shape = (1,), dtype = np.float32, minimum = 0, maximum = 1, name = 'action')**

**Comment by sguada** [comment]
Change your action spec to 
`array_spec.BoundedArraySpec(shape = (), dtype = np.float32, minimum = 0, maximum = 1, name = 'action')`

**Comment by ayarmak** [comment]
I'm getting a similar error, but I cannot apply the solution by @sguada above, because I need my action to be a tuple of two integers. 

My action_spec:
```
self._action_spec = array_spec.BoundedArraySpec(
        shape=(2,),
        dtype=np.int32, minimum=0, maximum=4, name='action')
```

The error I'm getting:
```
ValueError: Only scalar actions are supported now, but action spec is: BoundedTensorSpec(shape=(2,), dtype=tf.int32, name='action', minimum=array(0, dtype=int32), maximum=array(4, dtype=int32))
  In call to configurable 'DqnAgent' (<class 'tf_agents.agents.dqn.dqn_agent.DqnAgent'>)
```

Any advice on what I might be doing wrong?

**Comment by sguada** [comment]
As the error message shows, only scalar actions are supported for DQN, you can however transform your environment to have 1 action which represents the cross-product of all your actions.

**Comment by ayarmak** [comment]
This is super helpful, thanks a lot @sguada! Any chance you can share a bit more details or examples on how to potentially implement the workaround you mentioned? In my case, my action is a tuple of two numbers: 1) an id of a "source" category from which one item should be pulled and 2) an id of a "destination" category that this item would need to be re-assigned to. With those two numbers, my environment then updates the underlying state, reads an observation on that state, and calculates rewards. The environment does need both of those numbers though to implement the action. 

I assume what you mean is that there is a way to encode those two numbers as one, and then decode them back inside the environment (sorry if I misunderstood your point though). But I'm having some trouble imagining how to implement the "decoding" part...

**Comment by sguada** [comment]
Yeah you could manually convert the tuple of actions back and forth, or use

FlattenActionWrapper
https://github.com/tensorflow/agents/blob/master/tf_agents/environments/wrappers.py#L274

To do it automatically.

Sergio

**Comment by ayarmak** [comment]
Thanks a lot - this is super helpful @sguada !!

**Comment by jacklu333333** [comment]
Hi, @sguada 
I have the same error, but I am still a little bit confused. I try the automatic method with `FlattenActionWrapper`. My understanding is that `FlattenActionWrapper` will automatically flatten the action, and then I have to convert it to `TFPyEnvironment` so that `array_spec` will become `tensor_spec`.

However, when I tried the following code 

```
train_py_env = myenv()

train_env = tf_py_environment.TFPyEnvironment(
    wrappers.FlattenActionWrapper(
        train_py_env,
        np.uint8
    ), 
    isolation=True
)  
```
the following error pop up

`
only size-1 arrays can be converted to Python scalars
`
It seems the FlattenActionWrapper is ignored by `TFPyEnvironment`  somehow. I couldn't find any other example online.  I hope you can point out where did I do wrong.

Thanks in advance.

Best Regards,
Jack Lu

**Comment by KumarRupesh** [comment]
Hi, my action need to be a tuple of three integers.  I am using a  DQN agent and a custom PyEnvironment environment. I am also getting same error as Jack has shown above when I use  FlattenActionWrapper.  Please let me know if some one was able to resolve this.
